File: Hard/MonotonousQueue/1425_ConstrainedSubsequenceSum_MonoQueue.cpp
// Use mono queue to preserve max sum subsequence within a window of size k
// Time complexity: O(n)
// Space complexity: O(n)
const int INF = 0x3f3f3f3f;

class Solution {
public:
    int constrainedSubsetSum(vector<int>& nums, int k) {
        int n = nums.size();
        deque<int> q;
        int ans = -INF;
        vector<int> sum(n);
        for (int i = 0; i < n; i++)
        {
            sum[i] = nums[i];
            if (!q.empty()) sum[i] += sum[q.front()];
            ans = max(ans, sum[i]);
            while (!q.empty() && i - q.front() >= k) q.pop_front();
            while (!q.empty() && sum[q.back()] <= sum[i]) q.pop_back();
            if (sum[i] > 0) q.push_back(i);
        }
        return ans;
    }
};

File: Hard/MonotonousQueue/1425_ConstrainedSubsequenceSum_MonoQueue.py
class Solution:
    def constrainedSubsetSum(self, nums: List[int], k: int) -> int:
        q = collections.deque()
        n = len(nums)
        dp = [0] * n
        for i, v in enumerate(nums):
            dp[i] = v
            if len(q) > 0:
                dp[i] = max(dp[i], dp[i] + dp[q[0]])
            while len(q) > 0 and i - q[0] >= k:
                q.popleft()
            while len(q) > 0 and dp[i] >= dp[q[-1]]:
                q.pop()
            if dp[i] > 0:
                q.append(i)
        return max(dp)
            